Feat: Adds type annotations

peterschutt opened the pull request:
- Used MonkeyType and the test suite to generate  a set of annotations
- Added mypy and necessary stub dependencies to dev extras
- Adds types.py module for type definitions/aliases etc
- Adds py.typed
- Removes support for 3.7

Related to #39

peterschutt commented:
@tobymao you should be able to run `python -m mypy saq/queue.py` without error - I've focused my attention there so far and would like to give you the chance to weigh in before I go any further.

I'm running with default config to start with, we can ramp up the strictness settings of mypy once I'm past the initial batch of errors.

Found an issue with the redis stubs in typeshed: https://github.com/python/typeshed/issues/8960

tobymao commented on saq/job.py:
```diff
@@ -1,3 +1,5 @@
+
 import dataclasses
 import enum
 import typing
```
let's refactor everywhere that uses typing to the convention

import typing as t

tobymao commented on saq/job.py:
```diff
@@ -254,7 +259,7 @@ async def callback(_id, status):
             await self.refresh()
 
-    def replace(self, job):
+    def replace(self, job: "Job") -> None:
```
all types should be real types, not strings, let's just do Job here

barakalon commented on saq/queue.py:
```diff
@@ -20,20 +23,36 @@
+    from typing import Any, Callable, Dict, List, Optional, Set, Type, Union
+    from redis.asyncio.client import Redis
+    from redis.commands.core import AsyncScript
+    from .types import (
```
Should we avoid relative imports?

`from saq.types`

barakalon reviewed (approved):
yay!

Let's add mypy to `run_checks.sh`

tobymao commented on saq/types.py:
```diff
@@ -0,0 +1,30 @@
+if TYPE_CHECKING:
+    from typing import Any, Awaitable, Callable, Dict, Tuple, Union
+    from typing_extensions import TypeAlias
+    from .job import Job
```
remove relative imports

tobymao commented on saq/types.py:
```diff
@@ -0,0 +1,30 @@
+from __future__ import annotations
+
+from typing import TYPE_CHECKING, TypedDict
```
```suggestion
import typing as t
```

tobymao commented on saq/types.py:
```diff
@@ -0,0 +1,30 @@
+    from typing_extensions import TypeAlias
+    from .job import Job
+
+BeforeEnqueueType: TypeAlias = "Callable[[Job], Awaitable[Any]]"
```
don't use strings

peterschutt commented:
Thanks for the reviews! I'll keep working at it over the weekend.

> Let's add mypy to run_checks.sh

172 errors still when I run on saq and tests :) will def. add it in when I get a clean run.

tobymao commented:
let me know when is hould take another look

peterschutt commented:
A good chunk of the mypy errors revolve around attribute access on `Job.queue` as it might be `None`:

```
saq/job.py:117: error: Item "None" of "Optional[Queue]" has no attribute "name"  [union-attr]
saq/job.py:139: error: Item "None" of "Optional[Queue]" has no attribute "job_id"  [union-attr]
saq/job.py:217: error: Item "None" of "Optional[Queue]" has no attribute "abort"  [union-attr]
saq/job.py:221: error: Item "None" of "Optional[Queue]" has no attribute "finish"  [union-attr]
saq/job.py:225: error: Item "None" of "Optional[Queue]" has no attribute "retry"  [union-attr]
saq/job.py:235: error: Item "None" of "Optional[Queue]" has no attribute "update"  [union-attr]
saq/job.py:244: error: Item "None" of "Optional[Queue]" has no attribute "job"  [union-attr]
saq/job.py:257: error: Item "None" of "Optional[Queue]" has no attribute "listen"  [union-attr]
```
These are currently `AttributeError`s:

```python-traceback
Traceback (most recent call last):
  File "/home/peter/.pyenv/versions/3.10.8/lib/python3.10/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
  File "/home/peter/PycharmProjects/saq/saq/job.py", line 139, in id
    return self.queue.job_id(self.key)
AttributeError: 'NoneType' object has no attribute 'job_id'
```
One idea is to add an access method to centralise the exception logic and type narrowing, such as:

```python
class Job:
    ...
    def get_queue(self) -> Queue:
        if self.queue is None:
            raise TypeError("`Job` must be associated with a `Queue` before this operation can proceed")
        return self.queue

    @property
    def id(self) -> str:
        return self.get_queue().job_id(self.key)
```

It could also be a property and/or memoized but I doubt we'd be looking at too much overhead with the plain method.

Any opinion here @tobymao @barakalon?

tobymao commented:
> A good chunk of the mypy errors revolve around attribute access on `Job.queue` as it might be `None`:
> 
> ```
> saq/job.py:117: error: Item "None" of "Optional[Queue]" has no attribute "name"  [union-attr]
> saq/job.py:139: error: Item "None" of "Optional[Queue]" has no attribute "job_id"  [union-attr]
> saq/job.py:217: error: Item "None" of "Optional[Queue]" has no attribute "abort"  [union-attr]
> saq/job.py:221: error: Item "None" of "Optional[Queue]" has no attribute "finish"  [union-attr]
> saq/job.py:225: error: Item "None" of "Optional[Queue]" has no attribute "retry"  [union-attr]
> saq/job.py:235: error: Item "None" of "Optional[Queue]" has no attribute "update"  [union-attr]
> saq/job.py:244: error: Item "None" of "Optional[Queue]" has no attribute "job"  [union-attr]
> saq/job.py:257: error: Item "None" of "Optional[Queue]" has no attribute "listen"  [union-attr]
> ```
> 
> These are currently `AttributeError`s:
> 
> ```
> Traceback (most recent call last):
>   File "/home/peter/.pyenv/versions/3.10.8/lib/python3.10/code.py", line 90, in runcode
>     exec(code, self.locals)
>   File "<input>", line 1, in <module>
>   File "/home/peter/PycharmProjects/saq/saq/job.py", line 139, in id
>     return self.queue.job_id(self.key)
> AttributeError: 'NoneType' object has no attribute 'job_id'
> ```
> 
> One idea is to add an access method to centralise the exception logic and type narrowing, such as:
> 
> ```python
> class Job:
>     ...
>     def get_queue(self) -> Queue:
>         if self.queue is None:
>             raise TypeError("`Job` must be associated with a `Queue` before this operation can proceed")
>         return self.queue
> 
>     @property
>     def id(self) -> str:
>         return self.get_queue().job_id(self.key)
> ```
> 
> It could also be a property and/or memoized but I doubt we'd be looking at too much overhead with the plain method.
> 
> Any opinion here @tobymao @barakalon?

that’s fine

tobymao commented on saq/job.py:
```diff
@@ -233,7 +236,7 @@ async def update(self, **kwargs) -> None:
         await self.get_queue().update(self)
 
-    async def refresh(self, until_complete: t.Optional[float] = None) -> None:
+    async def refresh(self, until_complete: float | None = None) -> None:
```
is this syntax legal in 3.7?

peterschutt replied:
I was just about to write an explanation.

The `from __future__ import annotations` means that all annotations are strings at runtime, so no runtime errors irrespective of version, and type checkers are able to use the current syntax irrespective of the version.

I've been running the tests locally on 3.8 without issue and that syntax was introduced in 3.10. Similarly, the builtin subscription syntax, e.g., `list[...]` is 3.9+ but the future annotations mean we don't need to worry about the runtime errors.

peterschutt replied:
Of course, if there is an issue I'm missing here, or you'd prefer to stay syntactically valid with the min version irrespective of the future annotations, I'll be happy to roll them back. Just that the current syntax is a lot nicer than what was available in 3.7.

barakalon replied:
I think mypy will automatically add an Optional type when the default value is None? I could have that wrong, though.

peterschutt replied:
You are correct, but reliance on it is discouraged: https://github.com/python/peps/pull/689 and https://github.com/python/mypy/issues/9091

peterschutt commented on saq/types.py:
```diff
@@ -0,0 +1,43 @@
+    schedule: int
+    stats: int
+    sweep: int
+    abort: int
```
@tobymao Note in this module that I've used the typing versions of generics as it would be feasible that a user would import something from here to annotate their own code, and we can't be sure they'll use it with `from __future__ import annotations`, so needs to be compatible with 3.7 runtime.

This is also why `"Job"` and `"Status"` are forward refs here, as if this module is imported outside of a `if TYPE_CHECKING` block, their reference would otherwise fail.

tobymao replied:
can you double check everything works fine in 3.7?

peterschutt replied:
yes I will verify

peterschutt replied:
`IsolatedAsyncioTestCase` not available in 3.7. I tried to do a compat with [aiounittest](https://aiounittest.readthedocs.io/en/latest/info.html) for 3.7 (`import aiounittest.AsyncTestCase as `IsolatedAsyncioTestCase`) but there were a stack of errors. 

Perhaps a simple integration test script can be added to CI for 3.7 only?

3.7 EOL is 27th June 2023.

barakalon replied:
I've manually ported IsolatedAsyncioTestCase before. I can help with that.

@tobymao - is there some specific 3.7 use case you have in mind?

tobymao replied:
i guess we can leave saq at 3.8+, in my mind everything i support is 3.7+ because of airflow, but i guess there's no reason to mix airflow and saq

barakalon replied:
Then maybe we ignore 3.7?

tobymao replied:
yes

peterschutt replied:
that would certainly simplify things

peterschutt replied:
3.7 removed and added 3.11

peterschutt replied:
Note we still need to use the typing module generics here unless we go back to:

```python
LoadType: t.TypeAlias = "Callable[[bytes | str], t.Any]"
```
Six to one, half a dozen to the other in my mind, so I'll leave as is unless you'd prefer otherwise.

peterschutt commented on saq/types.py:
```diff
@@ -0,0 +1,43 @@
+CountKind = Literal["queued", "active", "incomplete"]
+DumpType = t.Callable[[t.Dict], str]
+DurationKind = Literal["process", "start", "total", "running"]
+Function = t.Callable
```
This is just a stub for a better type.

I tried different configurations with `Protocol` and `__call__` and `@overload` to try and satisfy requirement that these can be both sync and async, accept a single pos arg, either accept kwargs, or not, as well as having the `__qualname__` attribute, but couldn't find a solution that worked.

peterschutt replied:
Going 3.8+ might help with this as positional only parameters were introduced in 3.8.

peterschutt commented on saq/worker.py:
```diff
@@ -1,16 +1,28 @@
 
+if t.TYPE_CHECKING:
+    from asyncio import Task
+    from collections.abc import Callable, Collection, Coroutine
```
Are you happy for individual imports for `collections.abc`? Or prefer `from collections import abc` and `abc.Callable[...]` etc?

tobymao replied:
it’s ok

peterschutt commented:
I just added a commit that addresses issues with public interface types highlighted by the Pyright `--verifytypes` tool. For context, this was the output of the first run of the tool before any fixes.

Most of the issues were from use of generic types without specifying type arguments.

One interesting case to highlight was this:

```
saq.queue.Queue.__init__
  /home/peter/PycharmProjects/saq/saq/queue.py:72:9 - error: Type of parameter "redis" is partially unknown
    Parameter type is "Redis[Unknown]"
      Type argument 1 for class "Redis" has unknown type
```

`Redis` can be configured with `decode_responses=True` so that it returns `str` from commands, which would mean code like this would fail:

```python
        for key in await self.redis.zrangebyscore(self._stats, now(), "inf"):
            key_str = key.decode("utf-8")
```

Therefore `Queue` specifically expects `decode_responses=False` which is `Redis[bytes]`.

On the negative, it raised a few like this:

```
saq.worker.logger
  /home/peter/PycharmProjects/saq/saq/worker.py:27:1 - error: Type is missing type annotation and could be inferred differently by type checkers
    Inferred type is "Logger"

```

Meaning it wants:

```python
logger: logging.Logger = logging.getLogger("saq")
```

Pretty gross, and I'm not sure how that could be inferred to be anything different by other type checkers.

Full output:
```
(.venv) peter@peter-Inspiron-15-5510:~/test-saq$ pyright --verifytypes saq --ignoreexternal
Module name: "saq"
Package directory: "/home/peter/PycharmProjects/saq/saq"
Module directory: "/home/peter/PycharmProjects/saq/saq"
Path of py.typed file: "/home/peter/PycharmProjects/saq/saq/py.typed"

Public modules: 8
   saq
   saq.__main__
   saq.job
   saq.queue
   saq.types
   saq.utils
   saq.web
   saq.worker

Symbols used in public interface:
saq.job.Job.kwargs
  /home/peter/PycharmProjects/saq/saq/job.py:91:5 - error: Type argument 1 for class "dict" has unknown type
  /home/peter/PycharmProjects/saq/saq/job.py:91:5 - error: Type argument 2 for class "dict" has unknown type
saq.queue.Queue.__init__
  /home/peter/PycharmProjects/saq/saq/queue.py:72:9 - error: Type of parameter "redis" is partially unknown
    Parameter type is "Redis[Unknown]"
      Type argument 1 for class "Redis" has unknown type
saq.queue.Queue.redis
  /home/peter/PycharmProjects/saq/saq/queue.py:80:14 - error: Type argument 1 for class "Redis" has unknown type
saq.queue.Queue.uuid
  /home/peter/PycharmProjects/saq/saq/queue.py:82:14 - error: Type is missing type annotation and could be inferred differently by type checkers
    Inferred type is "str"
saq.queue.Queue.started
  /home/peter/PycharmProjects/saq/saq/queue.py:83:14 - error: Type is missing type annotation and could be inferred differently by type checkers
    Inferred type is "int"
saq.job.Job.meta
  /home/peter/PycharmProjects/saq/saq/job.py:110:5 - error: Type argument 1 for class "dict" has unknown type
  /home/peter/PycharmProjects/saq/saq/job.py:110:5 - error: Type argument 2 for class "dict" has unknown type
saq.job.Job.enqueue
  /home/peter/PycharmProjects/saq/saq/job.py:204:15 - error: Type of parameter "queue" is partially unknown
    Parameter type is "Queue | None"
saq.job.Job.get_queue
  /home/peter/PycharmProjects/saq/saq/job.py:266:9 - error: Return type is partially unknown
    Return type is "Queue"
saq.queue.logger
  /home/peter/PycharmProjects/saq/saq/queue.py:38:1 - error: Type is missing type annotation and could be inferred differently by type checkers
    Inferred type is "Logger"
saq.queue.JobError.__init__
  /home/peter/PycharmProjects/saq/saq/queue.py:44:9 - error: Type of parameter "job" is partially unknown
    Parameter type is "Job"
saq.types.QueueInfo.workers
  /home/peter/PycharmProjects/saq/saq/types.py:21:5 - error: Type argument 1 for class "dict" has unknown type
  /home/peter/PycharmProjects/saq/saq/types.py:21:5 - error: Type argument 2 for class "dict" has unknown type
saq.types.QueueInfo.jobs
  /home/peter/PycharmProjects/saq/saq/types.py:26:5 - error: Type argument 1 for class "list" has partially unknown type
    Type is dict[Unknown, Unknown]
saq.web.static
  /home/peter/PycharmProjects/saq/saq/web.py:23:1 - error: Type is missing type annotation and could be inferred differently by type checkers
    Inferred type is "str"
saq.worker.logger
  /home/peter/PycharmProjects/saq/saq/worker.py:27:1 - error: Type is missing type annotation and could be inferred differently by type checkers
    Inferred type is "Logger"
saq.worker.Worker.__init__
  /home/peter/PycharmProjects/saq/saq/worker.py:50:9 - error: Type of parameter "queue" is partially unknown
    Parameter type is "Queue"
  /home/peter/PycharmProjects/saq/saq/worker.py:50:9 - error: Type of parameter "functions" is partially unknown
    Parameter type is "Collection[((...) -> Unknown) | tuple[str, (...) -> Unknown]]"
      Type argument 1 for class "Collection" has partially unknown type
saq.worker.Worker.upkeep
  /home/peter/PycharmProjects/saq/saq/worker.py:161:15 - error: Return type is partially unknown
    Return type is "Coroutine[Any, Any, list[Task[Unknown]]]"
      Type argument 3 for class "Coroutine" has partially unknown type
saq.worker.Worker.event
  /home/peter/PycharmProjects/saq/saq/worker.py:111:18 - error: Type is missing type annotation and could be inferred differently by type checkers
    Inferred type is "Event"
saq.worker.Worker.functions
  /home/peter/PycharmProjects/saq/saq/worker.py:80:14 - error: Type is missing type annotation and could be inferred differently by type checkers
    Inferred type is "dict[Unknown, Unknown]"
  /home/peter/PycharmProjects/saq/saq/worker.py:80:14 - error: Type argument 1 for class "dict" has unknown type
  /home/peter/PycharmProjects/saq/saq/worker.py:80:14 - error: Type argument 2 for class "dict" has unknown type
saq.worker.Worker.cron_jobs
  /home/peter/PycharmProjects/saq/saq/worker.py:81:14 - error: Type is missing type annotation and could be inferred differently by type checkers
    Inferred type is "Collection[CronJob] | list[CronJob]"
saq.worker.Worker.tasks
  /home/peter/PycharmProjects/saq/saq/worker.py:83:14 - error: Type argument 1 for class "set" has partially unknown type
    Type is Task[Unknown]
saq.worker.Worker.job_task_contexts
  /home/peter/PycharmProjects/saq/saq/worker.py:84:14 - error: Type argument 2 for class "dict" has partially unknown type
    Type is dict[Unknown, Unknown]
saq.worker.async_check_health
  /home/peter/PycharmProjects/saq/saq/worker.py:332:11 - error: Type of parameter "queue" is partially unknown
    Parameter type is "Queue"

Symbols exported by "saq": 203
  With known type: 173
  With ambiguous type: 7
  With unknown type: 23
    (Ignoring unknown types imported from other packages)
  Functions without docstring: 50
  Functions without default param: 0
  Classes without docstring: 5

Other symbols referenced but not exported by "saq": 0
  With known type: 0
  With ambiguous type: 0
  With unknown type: 0

Type completeness score: 85.2%

Completed in 0.874sec
```

tobymao commented:
seems reasonable to me, is this ready to go?

peterschutt commented:
> seems reasonable to me, is this ready to go?

Close, I'm experimenting with it downstream. The top-level import aren't being recognized by mypy, e.g.,:

```python
import saq

q = saq.Queue()   # Module has no attribute "Queue"  [attr-defined]
```

Might need an `__all__` but I'm looking into it.

peterschutt commented:
> > seems reasonable to me, is this ready to go?
> 
> Close, I'm experimenting with it downstream. The top-level import aren't being recognized by mypy, e.g.,:
> 
> ```python
> import saq
> 
> q = saq.Queue()   # Module has no attribute "Queue"  [attr-defined]
> ```
> 
> Might need an `__all__` but I'm looking into it.

That was it ([ref](https://mypy.readthedocs.io/en/stable/command_line.html#cmdoption-mypy-no-implicit-reexport)).

Should be good to go now.

tobymao commented:
@peterschutt thanks again for this, great work!

peterschutt commented:
> @peterschutt thanks again for this, great work!

No worries! Feel free to ping me if there are any issues.
